src/main/java/com/conveyal/datatools/manager/utils/sql/TableCheck.java

```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Contains the outcome of a table check (i.e. whether columns are missing or of the wrong type).
+ */
+public class TableCheck {
+    public final String namespace;
+    public final Table table;
+    public final List<ColumnCheck> columns;
+    public final List<ColumnCheck> missingColumns = new ArrayList<>();
+    public final List<ColumnCheck> columnsWithWrongType = new ArrayList<>();
+
+    public TableCheck(Table table, String namespace, List<ColumnCheck> columns) {
+        this.namespace = namespace;
+        this.table = table;
+        this.columns = columns;
+
+        for (Field field : table.fields) {
+            Optional<ColumnCheck> foundColumnForField = columns
+                .stream()
+                .filter(c -> c.columnName.equals(field.name))
+                .findFirst();
+
+            if (foundColumnForField.isPresent()) {
+                ColumnCheck columnForField = foundColumnForField.get();
+                // Are fields that are present of the correct type?
+                if (!columnForField.isSameTypeAs(field)) {
+                    columnForField.setExpectedType(field.getSqlTypeName());
+                    columnsWithWrongType.add(columnForField);
+                }
+                // Only the id column seems to be marked as not nullable, so we won't check that for now.
+            } else {
+                missingColumns.add(new ColumnCheck(field));
+            }
+        }
+    }
+
+    public TableCheck(Table table, String namespace, SqlSchemaUpdater schemaUpdater) {
+        this(table, namespace, schemaUpdater.getColumns(namespace, table));
+    }
+
+    public boolean hasColumnIssues() {
+        return !missingColumns.isEmpty() || !columnsWithWrongType.isEmpty();
+    }
+
+    /**
+     * Builds the SQL statement to upgrade the table,
+     */
+    public String getAlterTableSql() {
+        return String.format(
+            "ALTER TABLE %s.%s %s;",
+            namespace,
+            table.name,
+            getColumnSql()
+        );
+    }
+
+    /**
+     * Builds the part of an SQL statement that adds or updates columns.
+     */
+    private List<String> getColumnSqlParts(
+        List<ColumnCheck> columns,
+        Function<ColumnCheck, String> mapper
+    ) {
+        return columns
+            .stream()
+            .map(mapper)
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Builds the part of an SQL statement that adds and/or updates columns.
+     */
+    private String getColumnSql() {
+        List<String> columnSqlParts = new ArrayList<>();
+        columnSqlParts.addAll(getColumnSqlParts(missingColumns, ColumnCheck::getAddColumnSql));
+        columnSqlParts.addAll(getColumnSqlParts(columnsWithWrongType, ColumnCheck::getAlterColumnTypeSql));
+        return String.join(", ", columnSqlParts);
+    }
+
+    public void printReport() {
+        if (hasColumnIssues()) {
+            System.out.println("          Issues in table: " + table.name);
```

Comment:
Consider tabs instead of spaces and printf instead of println for System outs.

Reply:
printf and `\t` are in place in https://github.com/ibi-group/datatools-server/pull/459/commits/1077f562e02bb6befd096535865e75f69e36b04c.